internal/nearcache/stale_read_detector.go

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nearcache
+
+import (
+	icluster "github.com/hazelcast/hazelcast-go-client/internal/cluster"
+	"github.com/hazelcast/hazelcast-go-client/internal/serialization"
+)
+
+type StaleReadDetector struct {
+	rh RepairingHandler
+	ps *icluster.PartitionService
+}
+
+func NewStaleReadDetector(rh RepairingHandler, ps *icluster.PartitionService) StaleReadDetector {
+	return StaleReadDetector{
+		rh: rh,
+		ps: ps,
+	}
+}
+
+func (sr StaleReadDetector) IsStaleRead(rec *Record) bool {
+	// port of: com.hazelcast.internal.nearcache.impl.invalidation.StaleReadDetectorImpl#isStaleRead
+	// key param in the original implementation is not used.
+	md := sr.rh.GetMetaDataContainer(rec.PartitionID())
+	return rec.UUID() != md.UUID() || rec.InvalidationSequence() < md.StaleSequence()
```

Comment:
uuid fields are atomic.Value right now in the go client. if there are no values stored in both uuid fields of `rec` and `md`,  `rec.UUID() != md.UUID()` will be false in go client, however `!rec.hasSameUuid(md)` in Java client will return true because it checks for null: https://github.com/hazelcast/hazelcast/blob/a0604ffa968a5d301b0b7c5adf1d1c8b887ac65b/hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/record/AbstractNearCacheRecord.java#L169

So I guess we should fix this behaviour difference.